### PR TITLE
Update jsem proposal to address problems discovered in GHC #25087 (under review)

### PR DESCRIPTION
**Pull request by wz1000**
The accepted [jsem proposal](https://github.com/ghc-proposals/ghc-proposals/pull/540) proposed a way to provide a paralellism control mechanism for GHC and build systems calling GHC using the system semaphore implmentation.

However, it was discovered in [GHC #25087](https://gitlab.haskell.org/ghc/ghc/-/issues/25087) that the system semaphore mechanism is not reliable when mixing jobserver and jobclient binaries compiled against different system C library implementations. Each libc implementation (glibc, musl etc.) is free to define its own shared memory structures to implement POSIX semaphore functionality, so as a consequence if the jobserver (i.e. cabal) and jobclient (i.e. ghc) are compiled using different `libc`s, they will disagree on the shape and usage of these in-memory data structures, resulting in a crash in the best case.

This is an attempt to update the language of the proposal to allow mechanisms which are not based on the system C library to provide this semaphore-based parallelism control mechanism.

It is paired with a re-implementation of the semaphore-compat library on POSIX systems, based on a simple protocol based POSIX domain sockets. However, the details of this implementation are not discussed in this proposal, instead the language has been broadened to allow semaphore-compat freedom to dictate its own protocol and mechanism, as long as the compatibility/versioning requirements laid out in the updated proposal are adhered to.

Care has been taken to ensure backwards/forwards compatibility and graceful degradation wherever possible (which it is in most cases).

semaphore-compat v2 changes: [gitlab.haskell.org/ghc/semaphore-compat/-/merge_requests/8](https://gitlab.haskell.org/ghc/semaphore-compat/-/merge_requests/8)

GHC MR: [gitlab.haskell.org/ghc/ghc/-/merge_requests/15729](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/15729)

Cabal MR: https://github.com/haskell/cabal/pull/11628

### CODE REVIEW

File: proposals/0540-jsem.rst [hunk 1]
```diff
@@ -171,18 +175,26 @@ know anything about the jobserver that has spawned them.
...
+must do so by means of this library.
 
 The implementation consists of two separate parts: jobservers and jobclients.
 We want GHC to act as a jobclient and ``cabal``/``stack`` to be jobservers.
 
+The interface supports the following operations:
+
+1. The jobserver must be able to create and destroy semaphores and initialise them with a
+  given number of tokens
+2. Clients must be able to connect to the semaphore with identifier (string) provided by the server
```
Comment (sheaf) [inline proposals/0540-jsem.rst:189]:
```suggestion
2. Clients must be able to connect to the semaphore with an identifier (a string) provided by the server
```

File: proposals/0540-jsem.rst [hunk 2]
```diff
@@ -246,6 +258,80 @@ with the semaphore: instead of systematically releasing a token to the semaphore
...
+a musl libc. Each libc implementation has its own independent semaphore implementation
+based on shared memory datastructures, and these are not compatible with each
+other. This has manifested in bug reports against ``cabal-install``
+([20]_) and GHC ([21]_).
+
+To support this method of distribution it becomes necessary to avoid using
+shared memory datastructures provided by the system C library to implement
+semaphores, and instead rely on some mechanism which is robust against mixing
+of jobservers and jobclients compiled against different system C library
+implementations.
```
Comment (sheaf) [inline proposals/0540-jsem.rst:290]:
```suggestion
To support this method of distribution, we must avoid using shared memory datastructures
provided by the system C library to implement semaphores, and instead rely on some
mechanism which is robust against mixing of jobservers and jobclients compiled against
different system C library implementations. 
```

File: proposals/0540-jsem.rst [hunk 3]
```diff
@@ -246,6 +258,80 @@ with the semaphore: instead of systematically releasing a token to the semaphore
...
+
+Currently, GHC 9.8.{1,2} and GHC 9.10.1 provide jobclient implementations based
+on ``semaphore-compat-1.0.0``, which ``cabal-install-3.12.1.0`` also uses to provide
+a jobserver implementation. ``semaphore-compat-1.0.0`` uses the problematic POSIX
+system semaphores and is susceptible to the restrictions regarding system C library
+usage by jobservers and jobclients.
+
+Since the underlying semaphore implementation is being fundamentally changed, the
+future version of ``semaphore-compat``, presumably shipped with a future version
+of GHC and ``cabal-install`` would be incompatible with existing jobservers and
```
Comment (sheaf) [inline proposals/0540-jsem.rst:318]:
```suggestion
future version of ``semaphore-compat`` – presumably shipped with a future version
of GHC and ``cabal-install`` – would be incompatible with existing jobservers and
```

File: proposals/0540-jsem.rst [hunk 4]
```diff
@@ -246,6 +258,80 @@ with the semaphore: instead of systematically releasing a token to the semaphore
...
+
+For instance, this means
+
+- ``cabal-install-3.12.1.0`` would be unable to compile code with the
+  ``--semaphore`` flag enabled for future versions of GHC shipping with the
+  updated ``semaphore-compat`` release. There could be a mechanism
+  in the new version of ``semaphore-compat`` (linked to by the future version of
+  GHC) to detect such a scenario and output an informative error message.
+
+- Similarly future versions of ``cabal-install`` would be unable to use
```
Comment (sheaf) [inline proposals/0540-jsem.rst:329]:
```suggestion
- Similarly, future versions of ``cabal-install`` would be unable to use
```

**Review by sheaf — APPROVED** [review]

**Comment by wz1000** [comment]
As there has been not much discussion, could we get a vote on this amendment?

**Comment by gridaphobe** [comment]
I have recommended acceptance to the committee.

**Comment by sgraf812** [comment]
Just from readinig the proposal (and not having read any of the GHC issues), I find myself unconvinced that shared memory as the underlying interprocess communication medium is the issue; it rather seems that the *breaking change in memory representation* between libc versions was. It's a bit unclear how using UNIX domain sockets makes this problem any better. Clearly, the bytes sent over the socket are susceptible to breaking changes in the serialised memory representation just as much as a shared memory mechanism. 

However, by *controlling* this memory representation in `semaphore-compat`, it seems that you can provide much better compat guarantees across different libc versions. One thing I wonder, because it is isn't spelled out in the proposal:

**What if you need to make a breaking change to the memory representation/communication protocol in the future? Is your protocol design in `semaphore-compat` forward compatible?**

For example, can `cabal-install-4.0` using `semaphore-compat-3.0.0` with a brand new protocol (or a Win32 API style backward compatible extension of the old protocol) continue to work as job server for old versions of `ghc` that depend on `semaphore-compat-2.0.0`? If the answer is "no, maintenance cost/benefit ratio is bad" then that is still a statement I would like to see in the proposal.

**Comment by wz1000** [comment]
> it rather seems that the breaking change in memory representation between libc versions was. 

This is correct.

> However, by controlling this memory representation in semaphore-compat, it seems that you can provide much better compat guarantees across different libc versions. One thing I wonder, because it is isn't spelled out in the proposal:

Yes, the point of the amendment is to allow `semaphore-compat` to control the exact communication protocol/representation instead of specifying it in the proposal. One downside of this is that build tools written in other languages would have to go through the haskell library (either by linking directly against it or by calling a helper binary linked against it) to use this feature instead of being able to re-implement it themselves.

We also considered writing a shared memory implementation or adopting the semaphore implementation from one of the libcs (`musl` in particular as it is much simpler), but decided against that as a shared memory implementation is significantly more complex than one based on Unix sockets. These semaphores aren't interacted with very often, so the performance of shared-memory data-structures is less of a concern.

> What if you need to make a breaking change to the memory representation/communication protocol in the future? Is your protocol design in semaphore-compat forward compatible?

This is a good question, and I don't think we can guarantee that there will never be any breaking changes in the future. There would have to be a high barrier of justification for such a change though, given that such a change would lead to breakage and require new GHC and cabal-install releases, along with any other tools that decide to adopt this feature. 

I think a good policy could be that each major version of `semaphore-compat` will be backwards/forwards compatible within the same major version, and that breaking changes to the protocol via a major version bump would have to be well justified and go through the GHC proposals process.

If this sounds sensible to you, I can add it to the amendment.

**Comment by sgraf812** [comment]
Alright, so the answer is basically "we have no forward compatibility plans because we do not anticipate any breaking changes to the protocol in the future". Seems fair enough. Yes, please add something like this to the proposal, and perhaps point out that the memory representaiton caused the breaking change and not the choice of interprocess communication primitive.

**Comment by wz1000** [comment]
I have updated the proposal with the compatibility policy I mentioned.

File: proposals/0540-jsem.rst [hunk 5]
```diff
@@ -171,18 +175,26 @@ know anything about the jobserver that has spawned them.
...
 
-The implementation relies on an interface for system semaphores. For this, we
-propose to using the functionality provided by the ``unix`` and ``Win32``
-libraries. In practice, our implementation ships with a compatibility layer
-which provides an abstraction over those two libraries. This handles creation of
-fresh semaphores, acquisition, and release of tokens, including a mechanism for
-interrupting a wait on a semaphore when a token is no longer needed.
+The implementation is built on ``semaphore-compat``, a Haskell library that
+provides an abstract semaphore interface. Tools that wish to use this feature
+must do so by means of this library.
```
Comment (simonpj) [inline proposals/0540-jsem.rst:180]: So the API of the library is part of this proposal. Please can you lay out the API of the library, here?

File: proposals/0540-jsem.rst [hunk 6]
```diff
@@ -171,18 +175,26 @@ know anything about the jobserver that has spawned them.
 Implementation of the protocol
```
Comment (simonpj) [inline proposals/0540-jsem.rst:175]: This is more than a mere implementation matter.   The `semaphore-compat` library is part of the specification.   Maybe re-title this subsection?

File: proposals/0540-jsem.rst [hunk 7]
```diff
@@ -246,6 +258,99 @@ with the semaphore: instead of systematically releasing a token to the semaphore
 once a job is done, we can instead re-use this token if we have other pending
 jobs. This prioritises the compilation of a single unit.
 
+Abstract Semaphore Interface
```
Comment (simonpj) [inline proposals/0540-jsem.rst:261]: This subsection is at least partly about (a) history and (b) implementation details.   None of it is part of the proposed specification.

Maybe move 
* back-compt stuff to a section on back-compat (see [template](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0000-template.rst)) .  Including the current 2.6 
* implementation stuff  (incl the current 2.5) to a Section on Implementation.  (Proposals often have such a section.)

**Review by simonpj — COMMENTED** [review]
I don't have a well-informed opinion about the technical content here (I'm sure it's good), but I do have views about the presentation.

Remember this proposal may be read in 5 yrs time, without any knowledge of history etc, and certainly without thinking of this diff as a thing itself.  The finished proposal should stand on its own.

**Comment by sgraf812** [comment]
> I have updated the proposal with the compatibility policy I mentioned.

Thank you. Can you also say something about the *cost* of a breaking change? It appears to me that a breaking change to `semaphore-compat` shipped in a future Cabal release (e.g. 3.16) means that this version will not work as a jobserver for any GHC prior to this release (e.g. < 9.12). Is that right? How would GHC 9.10 detect that it is incompatible with the jobserver of Cabal 3.16?
Would it just crash because of the unexpected memory layout of some message? Would it also crash the jobserver?

Sorry to be annoying about this, however I cannot help but thinking that we need *some* kind of versioning policy, unless we are *really really sure* that we will never change the protocol again.

---

One solution could be to open the UNIX socket at a different port for `semphore-compat-3` and continue to offer the old interface at the port for `semaphore-compat-2`. If you could simply validate and acknowledge this solution in the proposal, I would be happy.

IMO, the protocol specification is an *absolutely crucial part* of this proposal; it is what ensures that future cabal versions continue to work as job server with past GHC versions. I think you want to say "the implementation does not matter; it is shared in `semaphore-compat`" but then you still need to say how to detect when different major versions of that library talk to another.

**Comment by wz1000** [comment]
> Thank you. Can you also say something about the _cost_ of a breaking change? It appears to me that a breaking change to `semaphore-compat` shipped in a future Cabal release (e.g. 3.16) means that this version will not work as a jobserver for any GHC prior to this release (e.g. < 9.12). Is that right? How would GHC 9.10 detect that it is incompatible with the jobserver of Cabal 3.16?

`cabal-install` 3.16 would simply refuse to enable `-jsem` if it detects it is compiling with GHC 9.10.1. This is similar to what happens if you try to use `--semaphore` with `cabal-install` 3.12 and GHC 9.6.

The other case is more interesting, what happens if GHC 9.12 linked against `semaphore-compat-2.0.0` is called using a semaphore created by `cabal-install` 3.12 linked against `semaphore-compat-1.0.0`. Detecting this situation is simple, my idea is to prefix the semaphore identifier with a version number, so existing semaphore identifiers like `cabal_semaphore_a` in `semaphore-compat-1.0.0` would become `v2-cabal_semaphore_a` in `semaphore-compat-2.0.0`. This way GHC 9.12 will be able to detect that it has been called with an incompatible semaphore.

What to do when we detect such a situation is also worth discussing. I think we should emit a warning in GHC and continue compiling with `-j1` (or whatever `-j` is specified in the dynflags). This would let the build continue with the same level of parallelism as if you did not enable `cabal build --semaphore` or tried to used `cabal build --semaphore -w ghc-9.6`.

We could also loudly complain and crash, but I don't think this is such a good idea.

While it is helpful to sound out and discuss implementation details here, I don't think it is a good idea to specify them exactly in the proposal. I would prefer to specify requirements instead.

I can amend this proposal specifiying:

1. When a jobclient (ghc) tries to open/connect to an existing semaphore, it must be able to detect if the semaphore was created with an incompatible version of `semaphore-compat`. In case an incompatibility is detected, the jobclient must continue compiling with a single job thread, which is safe because the client is started with an implicit token that it inherits.
2. It is the jobserver's (cabal) responsibility to not call jobclients such as GHC 9.10.1 which are not compatible with the semaphore implementation and also do not have any form of incompatibility detection.

Point 2 will become vestigial over time as all jobclients in the wild adopt `semaphore-compat` 2.0+ and have some form of incompatibility detection.

---------------------------

Another question worth discussing is backport policy for the semaphore-compat version bump. Do we backport this change to future patch releases of GHC 9.10 and 9.8?

There are arguments for both backporting and holding off:

1. If we don't backport these changes, then GHC 9.8 and 9.10 will be forever unable to use `-jsem` unless you also stick to an older version of `cabal-install`.
2. Backporting these changes could definitely be considered a mild breaking change, as patched releases of GHC 9.8 and GHC 9.12 would fall back to reduced parallelism if used with old/unpatched versions of `cabal-install`.

On balance, I think backporting it makes sense given the likely scenario that we will support GHC 9.10 for quite a while in the future and patched versions of `cabal-install` will only become more prevalent during this.

**Comment by sgraf812** [comment]
Thanks, that all makes a lot of sense.

> Detecting this situation is simple, my idea is to prefix the semaphore identifier with a version number

Indeed, that sounds like a compelling solution. Why not make it part of the spec? That is, "the semaphore identifier also encodes the version of the protocol used".

>  I would prefer to specify requirements instead.

That's fine, but IMO it is important that these requirements are discussed in the proposal.

> When a jobclient (ghc) tries to open/connect to an existing semaphore, it must be able to detect if the semaphore was created with an incompatible version of semaphore-compat.

You could just say "[...], it must check the protocol version encoded in the semaphore identifier for compatibility".

---

I agree that backporting makes sense. It should be easier to upgrade cabal to a recent version (and to upgrade from GHC 9.8.2 to 9.8.3) than to upgrade from GHC 9.8.2 to 9.12 in order to profit from a working jobserver implementation.

**Comment by wz1000** [comment]
@simonpj and @sgraf812 I have updated the proposal taking your comments into account.

File: proposals/0540-jsem.rst [hunk 8]
```diff
@@ -130,59 +130,86 @@ GHC Jobserver Protocol
...
-know anything about the jobserver that has spawned them.
+The **semaphore identifier** must begin a a prefix ``v<N>-`` where the ``<N>``
+specifies the ``semaphore-compat`` major version that this semaphore was created
+using. For instance, a semaphore created with ``semaphore-compat-2.0.0`` could
+be called ``v2-some-semaphore12345``.
+
+*Jobclients* that are instantiated with an incompatible semaphore (as detected
+by looking at the version of the identifier) must continue the build without
+requesting any additional tokens, using only the implicit token they inherited
+from their parent process.
```
Comment (sgraf812) [inline proposals/0540-jsem.rst:180]: Simple and to the point. Perfect!

File: proposals/0540-jsem.rst [hunk 9]
```diff
@@ -246,6 +273,110 @@ with the semaphore: instead of systematically releasing a token to the semaphore
...
+
+Breaking changes to the protocol require a major version bump and need to be
+justified to and accepted by the GHC Proposals Committee.
+
+To allow breaking changes to the protocol to be detected we require that,
+starting with ``semaphore-compat-2.0.0``, the **semaphore identifier** must
+begin a a prefix ``v<N>-`` where the ``<N>`` specifies the ``semaphore-compat``
+major version that this semaphore was created using. For instance,
+a semaphore created with ``semaphore-compat-2.0.0`` could be called
+``v2-some-semaphore12345``.
```
Comment (sgraf812) [inline proposals/0540-jsem.rst:374]: Redundant with the specification? Generally I could live with a lot less text in this section, because the compat story is clear from the specification now.

**Review by sgraf812 — COMMENTED** [review]
LGTM! I wouldn't mind cutting down the section related to back compat, though. They are IMO redundant with the much improved specification. Perhaps keep the Cabal 3.12 and GHC 9.8 examples.

File: proposals/0540-jsem.rst (hunk 5 above)
Comment (simonpj) [reply proposals/0540-jsem.rst:193]: > So the API of the library is part of this proposal. Please can you lay out the API of the library, here?

You did not respond.... yet authors *must* use this API so it seems important to the proposal.  Am I off target?  This is far from what I know!

File: proposals/0540-jsem.rst [hunk 10]
```diff
@@ -246,6 +273,100 @@ with the semaphore: instead of systematically releasing a token to the semaphore
...
+
+As such, we must strive to avoid such breaking changes as much as possible.
+In the (hopefully rare) case when such a change cannot be avoided, the change
+must meet a high bar of justification.
+
+Concretely, the communication protocol used in each major version series
+of ``semaphore-compat`` must be compatible with other versions of the
+library belonging to the same major version series. For instance, a
+``semaphore-compat-2.0.0`` must be compatible with ``semaphore-compat-2.9.9``, but
+may not be compatible with ``semaphore-compat-3.0.0``.
```
Comment (sgraf812) [inline proposals/0540-jsem.rst:364]: This is not strictly relevant to the proposal, but do note that the [PVP](https://pvp.haskell.org/) implies that `semaphore-compat-2.1` is already a major bump from `2.0`. It would suffice to release the new version of `semaphore-compat` as 1.1.

File: proposals/0540-jsem.rst [hunk 11]
```diff
@@ -130,59 +130,86 @@ GHC Jobserver Protocol
...
   Each jobclient **must** release exactly as many tokens as it has acquired from
   the semaphore (this does not include the implicit token).
 
-All communication between the jobserver and the jobclient happens through the
-semaphore. This ensures modularity of the protocol: jobclients don't need to
-know anything about the jobserver that has spawned them.
+The **semaphore identifier** must begin a a prefix ``v<N>-`` where the ``<N>``
+specifies the ``semaphore-compat`` major version that this semaphore was created
+using. For instance, a semaphore created with ``semaphore-compat-2.0.0`` could
+be called ``v2-some-semaphore12345``.
```
Comment (sgraf812) [inline proposals/0540-jsem.rst:175]:
```suggestion
be called ``v2.0-some-semaphore12345``.
```

`2.0` is the major version according to the [PVP](https://pvp.haskell.org/). Of course you could choose to always bump the first component here, but I think it would not hurt to clarify.

**Review by sgraf812 — COMMENTED** [review]
Some PVP-related version confusions. Otherwise looks good to me, although I did not give it a full read.

File: proposals/0540-jsem.rst (hunk 11 above)
Comment (gridaphobe) [reply proposals/0540-jsem.rst:175]: I think it's fine, maybe even good, to distinguish between the *binary* compatibility of the semaphore implementation and the *source* compatibility of `semaphore-compat`. PVP is about source compatibility, but the authors of `semaphore-compat` might need to make a breaking API change while still preserving binary compatibility with previous versions of the library.

Using the first component of the package version to indicate binary compatibility level seems reasonable to me.

File: proposals/0540-jsem.rst (hunk 10 above)
Comment (gridaphobe) [reply proposals/0540-jsem.rst:364]: See my comment above about API vs (effectively) ABI changes. I think this lens makes sense to apply here, though I would rephrase to use PVP versions with the typical four components.

File: proposals/0540-jsem.rst (hunk 5 above)
Comment (gridaphobe) [reply proposals/0540-jsem.rst:193]: @wz1000 can you add the core operations here (i.e. `create`, `acquire`, `release`) and link to https://hackage.haskell.org/package/semaphore-compat-1.0.0/docs/System-Semaphore.html for the details.

Ultimately the library may evolve and the API Docs should be authoritative, but it's good to capture the starting point here for posterity.

**Comment by gridaphobe** [comment]
@wz1000 just a couple small pieces to address from @simonpj and @sgraf812 and then I think we should be good to merge.

**Comment by gridaphobe** [comment]
@wz1000 ping here, if you can address the remaining comments we can merge this proposal.

**Comment by wz1000** [comment]
I have updated the proposal addressing all the outstanding comments, and some new changes to match the implementation at https://gitlab.haskell.org/ghc/semaphore-compat/-/merge_requests/8

See also:

GHC MR: https://gitlab.haskell.org/ghc/ghc/-/merge_requests/15729

Cabal MR: https://github.com/haskell/cabal/pull/11628

File: proposals/0540-jsem.rst [hunk 12]
```diff
@@ -130,67 +134,117 @@ GHC Jobserver Protocol
...
+is a positive integer specifying the **protocol version** of the semaphore.
+For instance, a semaphore created with ``semaphore-compat-2.x`` uses protocol
+version 2 and could be called ``v2-some-semaphore12345``.
+
+Note that the protocol version is distinct from the PVP package version of
+``semaphore-compat``.  All releases within a major version series of
+``semaphore-compat`` (e.g. 2.0.0 through 2.9.9) **must** use the same
+protocol version and be wire-compatible with each other.  A protocol version
+bump (and corresponding major version bump of the library) requires approval
+from the GHC Proposals Committee.
```
Comment (adamgundry) [inline proposals/0540-jsem.rst]:
```suggestion
from the GHC Steering Committee.
```

**Review by adamgundry — COMMENTED** [review]
@simonmar please could you assign a new shepherd to this proposal? It looks good from my perspective, and it'll be good to get the `-jsem` feature working more reliably.

**Comment by simonmar** [comment]
Assigning myself as shepherd.

**Comment by simonmar** [comment]
Shepherd recommends acceptance, committee will review.

**Comment by wz1000** [comment]
There may be a few changes to the API for `semaphore-compat` arising from review, I will have the proposal updated by tomorrow. No other changes to the other text of the proposal though.

**Comment by Tritlo** [comment]
This v1 v2 business is not pretty, but I suppose it’s unavoidable. I wish we could default to the latest version instead of v1, and users could add v1 for backwards compatibility instead. Otherwise we'll have always have some users on v1, even if we're on v3 or v10. Is the v1 defaulting unavoidable? What would the impact be?

**Comment by sgraf812** [comment]
I understand the compat matrix such that newer Cabal releases supporting v2 will talk to older GHCs supporting only v1 by defaulting to `-jN`, circumventing the jobserver implementation entirely, no? v1 will only be picked by old Cabal + old GHC and that's pretty much unavoidable. v2 is picked only if both GHC and Cabal support it, which makes sense. Anyway, +1 from me.

**Comment by Tritlo** [comment]
as I understand it, you have to preface the semaphore identity with "v2-..." to use v2, is that wrong?

**Comment by adamgundry** [comment]
> as I understand it, you have to preface the semaphore identity with "v2-..." to use v2, is that wrong?

Well, I think that's right, but "you" here is typically Cabal or some other build tool (which takes care of semaphore identifier naming). Most end users will simply pass the boolean `--semaphore` option to Cabal (or perhaps later it can be turned on by default, and they won't have to do anything at all). So users will typically only "remain on v1" if they are stuck on an older GHC and Cabal version, which we can't do much about.

It looks to me like the version compatibility story is pretty well thought through. I think we should accept (and hopefully get this into 10.0).

**Comment by wz1000** [comment]
In answer to @Tritlo's comment, Adam and Sebastian are correct, cabal will pick the right version of the semaphore library itself.

I have updated the API with the new changes.

Note that on windows, now we stay at semaphore protocol version v1, this handles future potential protocol breakages a bit better.
